src/HttpServer.php

```diff
@@ -60,50 +65,104 @@ public function addMiddleware($middleware)
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
-                }
+            if ($request != null) {
+                $connection->removeAllListeners('data');
+                $connection->on('data', function ($data) use (&$connection, $that, &$request, &$transferredLength, &$input) {
+                    $that->handleBody($request, $data, $connection, $transferredLength, $input);
+                });
+                $connection->emit('data', array($data));
             }
         });
     }
 
+    /** @internal */
+    public function handleBody(RequestInterface $request, $data, ConnectionInterface $connection, &$transferredLength, $input)
+    {
+        if ($this->isChunkedEncodingActive($request)) {
+            $input->emit('data', array($data));
+            return;
+        }
+
+        if (!$request->hasHeader('Content-Length') && $data === '') {
+            // Simple request without body and without 'Content-Length'
+            $input->emit('end', array());
+            return;
+        }
+
+        if (!$request->hasHeader('Content-Length')) {
+            // 'Content-Length' is missing
+            $this->sendResponse(new Response(411), $connection);
+            return;
+        }
+
+        $contentLength = $request->getHeaderLine('Content-Length');
+
+        $int = (int) $contentLength;
+        if ((string)$int !== (string)$contentLength) {
+            // Send 400 status code if the value of 'Content-Length' is not an integer
+            $this->sendResponse(new Response(400), $connection);
+            return;
+        }
+
+        if (($transferredLength + strlen($data)) > $contentLength) {
+            // Only emit data until the value of 'Content-Length' is reached, the rest will be ignored
+            $data = substr($data, 0, $contentLength - $transferredLength);
+        }
+
+        $transferredLength += strlen($data);
+        $input->emit('data', array($data));
+
+        if ((int)$transferredLength === (int)$contentLength) {
+            // 'Content-Length' reached, stream will end
+            $input->emit('end', array());
+        }
+    }
+
+    /** @internal */
+    public function sendHttpBodyStream(RequestInterface $request, ConnectionInterface $connection)
```

Comment:
What does this method do? The method signature suggests it sends an outgoing request to the connection?!